python/hail/typecheck.py

```diff
@@ -0,0 +1,123 @@
+from decorator import decorator, getargspec
+from types import ClassType
+import re
+
+
+def extract(t):
+    m = re.match("<(type|class) '(.*)'>", str(t))
+    if m:
+        return m.groups()[1]
+    else:
+        return str(t)
+
+
+class TypeChecker(object):
+    def __init__(self, f, supported):
+        self.f = f
```

Comment:
For `pretty`, I would add a `nested` argument. That way, when things show up at the top-level there is no unnecessary parentheses. E.g.

```
dict of Any, Any

dict of Any, (dict of Any, Any)
```